src/Rules/Deprecations/EchoDeprecatedToStringRule.php

```diff
@@ -0,0 +1,115 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Deprecations;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt\Echo_;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\RuleLevelHelper;
+use PHPStan\Type\ErrorType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+
+/**
+ * @implements \PHPStan\Rules\Rule<Echo_>
+ */
+class EchoDeprecatedToStringRule implements \PHPStan\Rules\Rule
+{
+
+	/** @var RuleLevelHelper */
+	private $ruleLevelHelper;
+
+	public function __construct(RuleLevelHelper $ruleLevelHelper)
+	{
+		$this->ruleLevelHelper = $ruleLevelHelper;
+	}
+
+	public function getNodeType(): string
+	{
+		return Echo_::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if (DeprecatedScopeHelper::isScopeDeprecated($scope)) {
+			return [];
+		}
+
+		$messages = [];
+		foreach ($node->exprs as $key => $expr) {
+			$this->deepCheckExpr($expr, $scope, $messages);
+		}
+
+		return $messages;
+	}
+
+	/**
+	 * @param Node\Expr $expr
+	 * @param Scope $scope
+	 * @param string[] $messages
+	 */
+	private function deepCheckExpr(Node\Expr $expr, Scope $scope, array &$messages): void
+	{
+		if ($expr instanceof Node\Expr\BinaryOp) {
+			$this->deepCheckExpr($expr->left, $scope, $messages);
+			$this->deepCheckExpr($expr->right, $scope, $messages);
+		} elseif ($expr instanceof Node\Expr\Cast\String_ || $expr instanceof Node\Expr\Assign || $expr instanceof Node\Expr\AssignOp) {
+			$this->deepCheckExpr($expr->expr, $scope, $messages);
+		} elseif ($expr instanceof Node\Scalar\Encapsed) {
+			foreach ($expr->parts as $part) {
+				$this->deepCheckExpr($part, $scope, $messages);
+			}
+		} else {
+			$message = $this->checkExpr($expr, $scope);
+
+			if ($message) {
+				$messages[] = $message;
+			}
+		}
+	}
+
+	private function checkExpr(Node\Expr $expr, Scope $scope): ?string
+	{
+		$type = $this->ruleLevelHelper->findTypeToCheck(
```

Comment:
This core logic should be done differently. See how it's done here: https://github.com/phpstan/phpstan-deprecation-rules/blob/c13ab9a2ef62a0db0e866005cb522b5502bc9e38/src/Rules/Deprecations/CallToDeprecatedMethodRule.php#L41-L48